spmd/compiler/fusion.py

```diff
@@ -162,22 +214,448 @@ def _scan_graph_for_fusion_elements(
     return element_list
 
 
-def run_comm_fusion(gm: fx.GraphModule) -> bool:
+def _copy_fe_to_buffer(
```

Comment:
i'm pretty confused by this whole function.  it seems overly complicated.  why the need for separate tracing buffers and tracing a copy function?  i would try to just directly construct fx nodes that interact with my global buffer.

Reply:
The tracing allows me to leverage fx unrolling and making a series of nodes for what is otherwise a pretty complex series of node additions.  I would not want to do this by hand inserting nodes unless I'm missing something (though in thinking about this perhaps you mean taking the view/slice/copy nodes and then programmatically looping and adding everything that way?)

For context - Below is the pre and post nodes added by tracing and unrolling for a simple 2 call fusion:

<img width="1374" alt="annotated_final_graph" src="https://user-images.githubusercontent.com/46302957/204632778-88048ac2-f5a3-4a4c-9dd5-be6858cf0ad0.png">

In green highlight are the pre view/slice/copy nodes added in the pre aspect and after is the slice/view/copy nodes for moving back out.
Note that the buffer is a 1D buffer and thus we reshape in and out to handle any shapes seen.  If the buffer matched the dims then this would be a lot fewer nodes and could make more sense to just insert directly. 
I think though what you mean is to build a function that directly inserts nodes based on the known view/slice/copy for in and slice/view/copy for out instead of tracing and creating subgraphs?

Reply:
Actually I think you may have a good point - this could be done with a hybrid loop of node insertion to replicate the view/slice/copy and offset adjustments to replicate the loop unrolling via nodes and be done without the actual tracing. 

I've added it to the design doc and will test it out as a potential simplification. 
Thanks for the idea, gave you credit in the design doc.